types/react/experimental.d.ts

```diff
@@ -38,7 +38,24 @@ import React = require('.');
 
 export {};
 
+type NonNullableValue = string | number | boolean | object | symbol;
+
 declare module '.' {
+    type MutableSource<Source extends NonNullableValue, Version extends NonNullableValue> = Pick<{
```

Comment:
The type of `Version` seems to be overly broad despite React's own definitions, maybe it should be limited to `keyof any`.

Reply:
Version actually can be anything. It means a comparable snapshot of the source but is used internally by React, not the user.

Reply:
The user implements version types though right? They decide what a version should be, and the version points to a snapshot of the source. It probably won't be a boolean because it gives you only two versions, and I doubt people will be using non-primitives much due to referential integrity issues. If we're already going the route of restricting usage of non-nullable types, we might as well just restrict it even more, IMO.

It's easier to relax the types later than to restrict them after the fact

Reply:
Hmm, I've been thinking for a bit, but don't think we can restrict that more. I think a user can use a boolean version or even an observable boolean source.

createMutableSource is mainly used to track an external source that exists in a module or global scope. (except when it depends on context) So, in most cases, the source instance is static.

And the version is only used to detect tearing issues in Concurrent Mode: mutableSource determines if the version has changed when interruptible-rendering is resumed and decides whether to re-render the element.

Reply:
In that case, I would follow the same guidelines for restricting the type to be non-nullable (instead of using the big union, just use the generic type + non nullable parameter)

Reply:
> If we're already going the route of restricting usage of non-nullable types, we might as well just restrict it even more, IMO.

"We're" not doing it. React does. At least that is how I understand $NonMaybeType. I don't think we should be opinionated about what makes sense as a version. Especially if react allows booleans explicitly.

If we later restrict the type then only because react restricted it further.  Since this is still experimental I see no issue with release breaking changes more frequent.